File: code/binding/development/tools/cast/src/java/cast/architecture/WorkingMemoryChangeFilterComparator.java
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

/**
 * 
 */
package cast.architecture;

import java.util.Comparator;

import cast.cdl.FilterRestriction;
import cast.cdl.WorkingMemoryChange;
import cast.cdl.WorkingMemoryChangeFilter;
import cast.cdl.WorkingMemoryOperation;

public class WorkingMemoryChangeFilterComparator implements
		Comparator<WorkingMemoryChangeFilter> {

	private static WorkingMemoryChangeFilterComparator comparator;

	public static WorkingMemoryChangeFilterComparator getComparator() {
		if (comparator == null) {
			comparator = new WorkingMemoryChangeFilterComparator();
		}
		return comparator;
	}

	/**
	 * 
	 */
	private WorkingMemoryChangeFilterComparator() {
		// must use get method
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see java.util.Comparator#compare(java.lang.Object, java.lang.Object)
	 */
	public int compare(WorkingMemoryChangeFilter _f1,
			WorkingMemoryChangeFilter _f2) {

		// //System.out
		// .println("WorkingMemoryChangeFilterComparator.compare()");
		// //System.out.println(CAATUtils.toString(_f1));
		// //System.out.println(CAATUtils.toString(_f2));

		final int BEFORE = -1;
		final int EQUAL = 0;
		final int AFTER = 1;

		// same object
		if (_f1 == _f2) {
			return EQUAL;
		}

		
		// change id
		int comparison = _f1.address.id.compareTo(_f2.address.id);
		if (comparison != EQUAL) {
			return comparison;
		}

		// change sa
		comparison = _f1.address.subarchitecture
				.compareTo(_f2.address.subarchitecture);
		if (comparison != EQUAL) {
			return comparison;
		}

		// ontological type
		comparison = _f1.type.compareTo(_f2.type);
		if (comparison != EQUAL) {
			return comparison;
		}

		// operation
		if (_f1.operation.value() < _f2.operation.value()) {
			return BEFORE;
		}
		if (_f1.operation.value() > _f2.operation.value()) {
			return AFTER;
		}


		// xarch
		if ((_f1.restriction != FilterRestriction.LOCALSA)
				&& (_f2.restriction == FilterRestriction.LOCALSA)) {
			return BEFORE;
		}
		if ((_f1.restriction == FilterRestriction.LOCALSA)
				&& (_f2.restriction != FilterRestriction.LOCALSA)) {
			return AFTER;
		}

		// source of change
		comparison = _f1.src.compareTo(_f2.src);
		if (comparison != EQUAL) {
			return comparison;
		}

		// origin of filter
		comparison = _f1.origin.compareTo(_f2.origin);
		if (comparison != EQUAL) {
			return comparison;
		}

		
		// all comparisons have yielded equality
		// verify that compareTo is consistent with equals
		// (optional)
		// assert _o1.equals(_o2) : "compare inconsistent with
		// equals.";

		return EQUAL;
	}

	public static final boolean allowsChange(WorkingMemoryChangeFilter _filter,
			WorkingMemoryChange _change) {

		// System.out.println(CASTUtils.toString(_filter));
		// System.out.println(CASTUtils.toString(_change));

		// id options

		// if the filter id is not empty
		if (_filter.address.id.length() > 0) {
			// then ids much match
			if (!_filter.address.id.equals(_change.address.id)) {
				// System.out.println("id failed");
				return false;
			}

		}


		// subarchitecture options

		// if filter is set for a specific subarch
		if (_filter.address.subarchitecture.length() > 0) {
			// then subarchitectures must match
			if (!_filter.address.subarchitecture
					.equals(_change.address.subarchitecture)) {
				// System.out.println("sa failed");
				return false;
			}
		} else {
			// otherwises allow any subarchitecture in the change
		}

	

		// type options

		// if type set
		if (_filter.type.length() > 0) {
			// then they must match
			if (!_filter.type.equals(_change.type)) {
				// System.out.println("type failed");
				return false;
			}
		}

		
		// operation must be the same if not a wildcard
		if (_filter.operation != WorkingMemoryOperation.WILDCARD) {
			if (_filter.operation != _change.operation) {
				// System.out.println("operatopm failed");
				return false;
			}
		}
		
		// src options

		// if filter src set
		if (_filter.src.length() > 0) {
			// filter src must match
			if (!_filter.src.equals(_change.src)) {
				// System.out.println("src failed");
				return false;
			}

		}

		// if we get this far then we're fine

		return true;
	}

}

File: comsys/tags/release1/tools/cast/src/c++/cast/architecture/WorkingMemoryChangeFilterComparator.cpp
#include <cstring>
#include <iostream>

#include "WorkingMemoryChangeFilterComparator.hpp"

namespace cast {

  
  bool WorkingMemoryChangeFilterComparator::operator () (const cdl::WorkingMemoryChangeFilter & _f1, 
							 const cdl::WorkingMemoryChangeFilter & _f2) const {


    //used to try to keep mapping with identical Java code
    static const bool BEFORE = true;
    static const bool EQUAL = false;
    static const bool AFTER = false;

    // same object
    if (&_f1 == &_f2) {
      return EQUAL;
    }


    // change id
    int comparison = _f1.address.id.compare(
					    _f2.address.id);
    if (comparison < 0) {
      return BEFORE;
    }
    if (comparison > 0) {
      return AFTER;
    }
  
    // change sa
    comparison = _f1.address.subarchitecture.compare(
						     _f2.address.subarchitecture);
    if (comparison < 0) {
      return BEFORE;
    }
    if (comparison > 0) {
      return AFTER;
    }


    // ontological type
    comparison = _f1.type.compare(_f2.type);
    if (comparison < 0) {
      return BEFORE;
    }
    if (comparison > 0) {
      return AFTER;
    }

    // xarch
    if ((_f1.restriction != cdl::LOCALSA) && (_f2.restriction == cdl::LOCALSA)) {
      return BEFORE;
    }
    if ((_f1.restriction == cdl::LOCALSA) && (_f2.restriction != cdl::LOCALSA)) {
      return AFTER;
    }

    // operation
    if (_f1.operation < _f2.operation) {
      return BEFORE;
    }
    if (_f1.operation > _f2.operation) {
      return AFTER;
    }


    // source of change
    comparison = _f1.src.compare(
				 _f2.src);
    if (comparison < 0) {
      return BEFORE;
    }
    if (comparison > 0) {
      return AFTER;
    }

    // originator
    // change id
    comparison = _f1.origin.compare(
				    _f2.origin);
    if (comparison < 0) {
      return BEFORE;
    }
    if (comparison > 0) {
      return AFTER;
    }

  
    return EQUAL;
  
  }


  bool 
  WorkingMemoryChangeFilterComparator::allowsChange(const cdl::WorkingMemoryChangeFilter &_filter,
						    const cdl::WorkingMemoryChange &_change) {
  
    //std::cout<<"allowsChange: "<<_change.type<<std::endl;

    // id options
    
    // if the filter id is not empty
    if (_filter.address.id.length() != 0) {
      // then ids much match
      if (_filter.address.id != _change.address.id) {
	return false;
      }    
    }

    // subarchitecture options
    
    // if filter is set for a specific subarch
    if (_filter.address.subarchitecture.length() != 0) {      
      
      // then subarchitectures must match
      if (_filter.address.subarchitecture != _change.address.subarchitecture)  {
	//std::cout<<_filter.address.subarchitecture<<" == "<< _change.address.subarchitecture<<std::endl;
	return false;
      }
    }
    else {
      // otherwises allow any subarchitecture in the change
    }
    
    //std::cout<<"passed sa"<<std::endl;

    //type options

    //if type is set
    if (_filter.type.length() != 0) {      
      // then they must match directly or the filter type must be a member
      // of the supertypes of the change type
      
      bool allowsTypeChange = false;
      
      // direct comparison
      if (_filter.type == _change.type) {  
	//printf("allowing based on exact match: %s\n", _filter.type.c_str());
	allowsTypeChange = true;
      } 
      else {
	// sanity check to allow assumptions later
	//assert (_change.type
	//.equals(_change.superTypes[_change.superTypes.length - 1]));
	
	// go through the super types except for the most derived one
	// (which is _change.type)
	for (unsigned int i = 0; i < _change.superTypes.size() - 1; ++i) {
	  // if the filter matches against the supertype
	  if (_filter.type == _change.superTypes[i]) {
	    //printf("allowing based on supertypes: %s\n", _filter.type.c_str());
	    allowsTypeChange = true;
	    break;
	  }
	}
      }
      
      if (!allowsTypeChange) {
	  // printf("type failed");
	return false;
      }
      
    }
    
    // operation must be the same if not a wildcard
    if(_filter.operation != cdl::WILDCARD) {
      if (_filter.operation != _change.operation) {
	return false;
      }
    }
  
    //std::cout<<"passed op"<<std::endl;
    

    //std::cout<<"passed type"<<std::endl;

  
    //std::cout<<"passed src"<<std::endl;
      
    // src options
    
    // if filter src set
    if (_filter.src.length() != 0) {      
      // filter src must match
      if (_filter.src != _change.src) {
	return false;
      }
    
    }

    
    //std::cout<<"passed id"<<std::endl;

    // if we get this far then we're fine
    
    return true;
  }
    

} //namespace cast
